---
title: "Mapping to Orthogroups"
author: "Bin He"
date: "3/12/2020"
output: 
  html_document:
  toc: TRUE
  toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(tidyverse)
require(janitor) # for adding total row
```

## Questions

Did _C. auris_, _C. albicans_ and _C. glabrata_ inherit the same set of adhesin families? In other words, do they use an ancestral pool of adhesin genes or did they evolve species-specific ones by co-opting or _de novo_ evolution? If they share certain families, are the number of paralogs (members of a family) similar or different?

## OrthoMCL

I used the EuPathDB's [Galaxy site](http://eupathdb.globusgenomics.org/) to map the predicted adhesins from all five genomes to "orthogroups". These are pre-computed groups of putative orthologous protein sequences. These are calculated using the OrthoMCL program. Here is a [description](https://orthomcl.org/orthomcl/about.do#background) of how that is done. 

Briefly
- All-v-all BLASTP of the proteins.
- Compute percent match length
    - Select whichever is shorter, the query or subject sequence. Call that sequence S.
    - Count all amino acids in S that participate in any HSP (high scoring pairs, from BLAST local alignment).
    - Divide that count by the length of S and multiply by 100
- Apply thresholds to blast result. Keep matches with E-Value \< 1e-5 percent match length \>= 50%
- Find potential inparalog, ortholog and co-ortholog pairs using the Orthomcl Pairs program. (These are the pairs that are counted to form the Average % Connectivity statistic per group.)
- Use the MCL program to cluster the pairs into groups

## Data

OrthoMCL output, including `mappedGroups.txt`, `MCS.tabular` and `paralogPairs.txt`. Among them, the `mappedGroups.txt` contains the mapping between the proteins in the input file and the pre-calculated Orthogroups. `paralogPairs.txt` gives pairwise distance calculated among the input sequences. `MCS.tabular` contains the clustered proteins based on the `paralogPairs.txt`. Note these should only include those sequences that don't get mapped to the existing Orthogroups.

## FungalRV results

### Load data

```{r load_data}
# the following file contains the metadata for all genes
# including species, strains, protein_id and FungalRV score
# threhold used by current version of Fungal RV: Score > 0 

frv.res <- read_tsv("../../output/FungalRV/all-fungalrv-results-20200228.txt", comment = "#")

# the following file contains the mapped results
frv.mapped <- read_tsv("all-fungalRV-predicted-orthoMCL-mappedGroups-20200309.txt")
# merge with frv.res to provide meta information
frv.mapped <- frv.mapped %>% 
  left_join(frv.res, by = c("protein_id"="ID")) %>% 
  select(Species, Strain, protein_id, everything())

# the following file contains unmapped genes grouped by MCS algorithm
frv.mcs <- read_tsv("all-fungalRV-predicted-orthoMCL-MCS-20200309.txt", comment = "#")
# merge with frv.res to provide meta information
frv.mcs <- frv.mcs %>% 
  left_join(frv.res, by = c("protein_id"="ID")) %>% 
  select(Species, Strain, protein_id, everything())
```

### Examine OrthoMCL mapping results
I submitted all predicted adhesins in the five genomes to the Galaxy site. How many of them were successfully mapped for each species/strain?

```{r percent_mapped}
# total number of adhesins submitted per species
frv.res %>% 
  filter(Score > 0) %>% 
  count(Strain) %>% 
  adorn_totals()

# combined both mapped and MCS clustered

```
### Tally numbers and sort
My goal is to count the number of members in each species/strain that belong to an orthogroup, ignoring those with just one member, as they have to be species specific. Then report the subtotals ordered by the total number of members in each orthogroup.

```{r subtotal_orthogroup}
# next let's look at the mappings
faa.mapped.sub <- faa.mapped %>% 
  count(Strain, orthomcl_group_id) %>% # group by strain and orthogroup ID, then count
  spread(key = Strain, value = n)  # spread the Strain variable so that all counts per strain is in the same row

faa.mapped.tot <- faa.mapped %>% 
  count(orthomcl_group_id, name = "total") %>% 
  filter(total > 1) # remove orthogroups with only one member across all species/strains

faa.mapped.final <- faa.mapped.sub %>% 
  right_join(faa.mapped.tot) %>% # this joins the two table and only retails rows that exist in mapped.tot
  arrange(desc(total))

print(faa.mapped.final, n=Inf)
```

## FaaPred results

### Load data

```{r load_data_faapred}
frv.res <- read_tsv("../../output/FungalRV/all-fungalrv-results-20200228.txt", comment = "#")
faa.mapped <- read_tsv("all-faapred-predicted-orthoMCL-mappedGroups-20200309.txt")
faa.mapped <- faa.mapped %>% 
  select(-evalue_mantissa) %>% 
  left_join(frv.res, by = c("protein_id"="ID")) %>% 
  select(Species, Strain, protein_id, everything())
```

### Examine OrthoMCL mapping results
I submitted all predicted adhesins in the five genomes to the Galaxy site. How many of them were successfully mapped for each species/strain?

## Tally numbers and sort
My goal is to count the number of members in each species/strain that belong to an orthogroup, ignoring those with just one member, as they have to be species specific. Then report the subtotals ordered by the total number of members in each orthogroup.

```{r subtotal_orthogroup_faapred}

# next let's look at the mappings
faa.mapped.sub <- faa.mapped %>% 
  count(Strain, orthomcl_group_id) %>% # group by strain and orthogroup ID, then count
  spread(key = Strain, value = n)  # spread the Strain variable so that all counts per strain is in the same row

faa.mapped.tot <- faa.mapped %>% 
  count(orthomcl_group_id, name = "total") %>% 
  filter(total > 1) # remove orthogroups with only one member across all species/strains

faa.mapped.final <- faa.mapped.sub %>% 
  right_join(faa.mapped.tot) %>% # this joins the two table and only retails rows that exist in mapped.tot
  arrange(desc(total))

print(faa.mapped.final, n=Inf)
```