---
title: Protein family classification for predicted adhesins in all five genomes 
author: Bin He
date: 2020-03-12
---
# Summary
The goal of this analysis is to answer the question "what are the evolutionary relationships between the predicted adhesins in _C. auris_, _C. albicans_ and _C. glabrata_?". The reason for this quesiton is because we want to distinguish two hypotheses about the origin and evolution of adhesin genes:
1. Adhesin genes were ancestral and inherited by the descendent species, such as _C. albicans_ and _C. auris_ (_C. glabrata_ is more distantly related, but the same question applies).
1. Adhesin genes were mostly species-specific, meaning that they evolved in each species from non-adhesin genes, or _de novo_ from non-coding DNA.

# Content
Our approach to answer this question and the results are documented in the R markdown files. Below are the links to view their results as HTML:

| File | Description | Link | 
|------|-------------| ---- |
| OrthoMCL.Rmd | Compare OrthoMCL-v5 and v6r1 results, initial summary (replaced by the new analysis below) | https://htmlpreview.github.io/?https://github.com/binhe-lab/C037-Cand-auris-adhesin/blob/master/01-global-adhesin-prediction/analysis/protein-family-classification/OrthoMCL.nb.html |
| OrthoMCL2-v5.Rmd | Test new ideas, like limiting to the FaaPred predicted, and add _S. cerevisiae_ genes | https://htmlpreview.github.io/?https://github.com/binhe-lab/C037-Cand-auris-adhesin/blob/master/01-global-adhesin-prediction/analysis/protein-family-classification/OrthoMCL2-v5.nb.html |
| orthomcl-gene-exploration.Rmd | A shiny app for interactive exploration of the OrthoMCL result! | https://binhe-lab.shinyapps.io/orthomcl-gene-exploration/ |

Other files

| File | Description |
|------|-------------|
| S_cerevisiae_orthoMCL_v5_download.txt | Directly downloaded from [FungiDB Release 46](https://fungidb.org/fungidb) |
| orthomcl-v5-mapping-analysis-cache.RData | Saved image generated by OrthoMCL2-v5.Rmd, used by the shiny app |
| scer_id.txt | a list of ref_protein IDs for _S. cerevisiae_ genes |

# Notes
## 2020-03-12 [HB] Motivation for this analysis

_Question_

Did _C. auris_, _C. albicans_ and _C. glabrata_ inherit the same set of adhesin families? In other words, do they use an ancestral pool of adhesin genes or did they "come up with" species-specific adhesins, by co-opting or _de novo_ evolution? If they share certain families, are the number of paralogs (members of a family) similar or different?

_Approach_

Our goal is to map the predicted adhesins from all five genomes to "orthogroups", which are pre-computed groups of putative orthologous protein sequences. These are calculated using the OrthoMCL program. Here is a [description](https://orthomcl.org/orthomcl/about.do#background) of how that is done. Briefly

- All-v-all BLASTP of the proteins.
- Compute percent match length
    - Select whichever is shorter, the query or subject sequence. Call that sequence S.
    - Count all amino acids in S that participate in any HSP (high scoring pairs, from BLAST local alignment).
    - Divide that count by the length of S and multiply by 100
- Apply thresholds to blast result. Keep matches with E-Value \< 1e-5 percent match length \>= 50%
- Find potential inparalog, ortholog and co-ortholog pairs using the Orthomcl Pairs program. (These are the pairs that are counted to form the Average % Connectivity statistic per group.)
- Use the MCL program to cluster the pairs into groups

On 2020-03-10, I (Bin He) submitted all predicted adhesins to the workflow for mapping proteins to orthogroups, made availabe by EuPathDB's Galaxy site (see `output/OrthoMCL` for details). In this folder, I'm analyzing the output to answer the above questions.


## 03/16/2020 [RS] OrthoMCL Information

Algorithm information is [here](https://docs.google.com/document/d/1RB-SqCjBmcpNq-YbOYdFxotHGuU7RK_wqxqDAMjyP_w/pub) and protocol information is [here](https://currentprotocols.onlinelibrary.wiley.com/doi/full/10.1002/0471250953.bi0612s35)

## 2020-03-20 [HB] Re-ran OrthoMCL mapping for all fungalRV predicted adhesins

1. Concatenated all FungalRV predicted adhesins to a single fasta file
1. Upload onto [OrthoMCL galaxy site](https://eupathdb.globusgenomics.org/) and run the pipeline
1. There are two result files, the `MCS.tabular` and the `mappedGroups.txt`. The former contains sequences not mapped to any existing Orthogroups, and are further clustered using the MCS algorithm.
1. I manually edited the `MCS.tabular` so that each row is a protein sequence and the same cluster (at most three in a cluster) gets an arbitrary MCS_# number.
1. On Galaxy OrthoMCL pipeline, I varied the "inflation" parameter for the MCS program according to the program's suggestion. the default is 4, and I changed it 1.4 and 2.0  Neither changed the results.

## 2020-05-15 [HB] Learn how OrthoMCL-DB tool deal with proteins that have no hit to their precomputed groups
According to Fischer et al. 2011, under _How the tool processes your proteins_, the authors explained:

> _Phase 1_
> In this phase the tool maps your proteins to groups in OrthoMCL-DB. It performs a BLASTP against all the proteins in OrthoMCL-DB, using a cutoff of e-5 and 50% match. Your protein is assigned to the group containing its best hit. If the best matching protein does not have a group, it is assigned to NO_GROUP.
>
> _Phase 2_
> In this phase the tool processes all your proteins that did not have an above-threshold match in phase 1. It uses the OrthoMCL-DB in-paralog algorithm described in the OrthoMCL Algorithm Document to create pairs of potential in-paralogs. It then submits those pairs to the MCL program for clustering. The result is groups of your proteins that provisionally represent in-paralogs.

So basically, any proteins that don't have a hit in the OrthoMCL database will be subject to the MCL (Markov Clustering Algorithm) for identifying potential in-paralogs. It also tells us that the majority of the proteins that do have hits in the database are not even processed with the MCL algorithm.

## 2020-05-15 [HB] Re-process the FungalRV predicted adhesins using the VEuGalaxy's OrthoMCL workflow, but changing to v6r1 of the databased released in April 2020
By comparing the results from v5 and v6, I found that both version would leave some input sequences unmapped. This is largely because the OrthoMCL-DB v5 and v6r1 contain different number of sequences. The v6r1 is much larger, because it contains a lot more species. What's unclear to me is why some sequences can map to v5 but not v6r1 -- it seems some orthogroups present in v5 are dropped in v6r1.

## 2020-05-22 [HB] Thoughts and next set of questions with the OrthoMCL analyses
1. OrthoMCL is different from PFam or EC, which group genes based on the putative or demonstrated functions. Here genes are grouped based on evolutionary relationship, inferred from sequence similarity. That is to say, genes with similar functions, e.g. adhesins, may not belong to the same orthogroup. This is exactly what I intended, since my initial question is an evolutionary one: are adhesin genes in _C. auris_ belong to the same orthogroups as those in _C. albicans_ and _C. glabrata_? In other words, did all three species _inherit_ the same set of genes with adhesin functions, or did they each evolve new adhesin proteins that are evolutionarily _unrelated_ (meaning they didn't descend from the same ancestral gene or gene family in their common ancestor)?
1. In practice, the orthogroups are just guesses at the evolutionary relationship. The Markov Clustering Algorithm also has an inflation ratio parameter that defines how "tight" the clusters are. All these mean that the orthogroup definitions are somewhat arbitrary.
1. That being said, the OrthoMCL-DB is based on a very large number of proteomes (predicted for most species, I think) and thus is likely to capture a lot of the ancient families. Gene families that are more phylogenetically restricted are more likely to be missed, if the species in which they exist are not well sampled. This shouldn't be the case for fungi, and especially not a problem for the _Saccharomycotina_ order.
1. Under the above considerations, I think I have to take the evidence from the OrthoMCL analyses as saying most of the adhesins don't belong to the same orthogroups in the three species.

## 2020-05-28 [HB] Examine top orthogroups and gene functions
- I separated the gene annotation part from the `orthomcl.Rmd`, which include the exploratory analyses with two versions of OrthoMCL-DB, and some summary statistics of the mapped genes in the five proteomes.
- I made a new `orthomcl-gene-exploration.Rmd`, which uses a `shiny` app to allow interactive exploration of the gene functions. The annotations were fetched from the Conserved Domain Database based on the ref_protein IDs.
- I manually examined the top 10 orthogroups. The top 5 clearly make sense -- the CDD showed that members were annotated as cell wall proteins or adhesins. Some of the other orthogroups, however, don't seem to make a lot of sense, e.g. alcohol dehydrogenase, HSP12 heat shock protein. Quite a few of the orthogroups represent enzymes, such as the glucosidase, aspartyl protease and glycosyl hydrolases. One orthogroup, represented by CaKre9 that is likely involved in cell wall regulation.

## 2020-05-29 [HB] Next step questions
1. Some of the orthogroups based on all of the fungalRV prediction results seem unrelated to adhesion. How about the FaaPred filtered ones?
1. Can we use the orthoMCL results and other protein sequence features, such as CATH, GPI-anchor prediction, signal peptide prediction, chromosomal location (in telomeric regions?) and low complexity repeat regions.
1. Can we add _S. cerevisiae_ results and get a bit more clarity?

_todo_

1. add FaaPred (Y/N) to the gene table
1. add _S. cerevisiae_ to the gene table, including its orthogroup mapping
1. start a new Rmd by duplicating the previous one, use a sandbox folder?

## 2020-06-05 [HB] Todo list
1. Merge _S. cerevisiae_ results, add FaaPred flag, and repeat the orthomcl results
    - percent of predicted adhesins mapped
    - merge the three _C. auris_ adhesin sets
    - proportion of predicted adhesins belonging to an orthogroup that is shared by at least two genomes
1. Examine top orthogroup functions
1. Export the sequences belonging to an orthogroup and explore their evolutionary relationship.

## 2020-06-06 [HB] Continue with OrthoMCL analysis
1. Downloaded orthogroup and other gene information for the four species (for _C. auris_ only B8441 is provided) from FungiDB beta (more details see `../../output/OrthoMCL/README.md`
1. When examining gene members in a single orthogroup, I asked myself what does it mean when one species have multiple members in that orthogroup? Were they species specific duplicates or were they ancient duplications shared across several species? To answer this question, I tested one orthogroup, OG5_132054, by performing alignment and tree reconstruction. See `../../output/OrthoMCL/all-fungalRV-orthoMCL-v5/subset/README.md` for details.
## 2020-06-12 [HB] Observations based on the latest version of `orthomcl-gene-exploration.Rmd`
I finished updating the shinyapp to include all results based on OrthoMCL-DB v5. Four different subsets can be queried individually, corresponding to different selection criteria (liberal to stringent). Below are my summary of the observations:

### FungalRV stringent (Score > 0.511)
| Group_id   | Caur| Calb | Cgla | Scer | Putative function (based on Conserved Domain Database unless noted) | 
|------------|-----|------|------|------|---------------------------------------------------------------------| 
| OG5_126579 | 3   | 8    | 5    | 11   | ALS family                                                          | 
| OG5_132045 | 8   | 8    | 0    | 0    | Hyr1 family, hyphally regulated cell wall GPI-anchored protein 1    | 
| OG5_152943 | 0   | 0    | 14   | 0    | _C. glabrata_ specific, members contain domains such as<br>CAGL0J05159g: "Hyphally regulated cell wall protein N-terminal"<br>CAGL0I10362g: "bacterial beta-glucosidase", "GLEYA domain, lectin-like binding found in _S. cerevisiae_ Flo proteins" | 
| NO_GROUP   | 1   | 4    | 3    | 0    | _C. glabrata_ CAGL0E06600g: Hyphally regulated cell wall protein N-terminal, Flocculin repeat | 
| OG5_132246 | 0   | 0    | 5    | 1    | includes _S. cerevisiae_ Flo10, similar domains as the 3rd orthogroup above | 
| OG5_136169 | 1   | 1    | 1    | 2    | Beta-glucosidase, _C. albicans_ _SUN41_: Cell surface beta-glucosidase involved in cytokinesis, cell wall biogenesis, adhesion to host tissue, and biofilm formation [UniProt](https://www.uniprot.org/uniprot/Q59NP5) | 
| OG5_138681 | 1   | 1    | 1    | 1    | Glycosyl hydrolase, _C. albicans_ member has "cell wall organization" [UniProt](https://www.uniprot.org/uniprot/A0A1D8PNW1) | 
| OG5_145217 | 0   | 2    | 1    | 1    | Flocculin type 3 repeat, pfam13928, including _S. cerevisiae_ _CCW12_ | 
| OG5_126661 | 1   | 1    | 0    | 1    | Alcohol dehydrogenase GroES-like domain<sup>1</sup> | 
| OG5_159685 | 0   | 0    | 2    | 1    | _S. cerevisiae_ _DAN4_: component of the cell wall, extensively O-glycosylated, GPI-anchor [UniProt](https://www.uniprot.org/uniprot/P47179) | 
| OG5_160783 | 0   | 1    | 2    | 0    | _C. albicans_ _PGA62_: fungal cell wall organization, N- and O-glycosylated, GPI anchor [UniProt](https://www.uniprot.org/uniprot/Q5AF41)<br>_C. glabrata_ CAGL0L06424g: Flocculin type 3 repeat and many other domains | 
| OG5_194847 | 0   | 0    | 1    | 2    | _S. cerevisiae_ _SRL1_: required to stablize cell wall in the presence of multiple GPI-anchored mannoproteins [UniProt](https://www.uniprot.org/uniprot/Q08673) | 

_Notes_

1. Defying my naive expectation that a cytosolic enzyme such as alcohol dehydrogenase can have nothing to do with adhesion, I found [this review](https://mmbr.asm.org/content/72/3/495) and a [paper](https://pubmed.ncbi.nlm.nih.gov/11700367/) it cited that says otherwise. Below I quote from the Discussion section of the second paper:
    > The possibility that a cell-surface-associated cytosolic enzyme could serve in the capacity of an ECM protein receptor andÂ’or adhesin is well accepted in other micro- organisms. For example, E. histolytica ADH binds fibronectin, collagen type II and laminin (Yang et al., 1994). Along similar lines, the fibronectin receptor of Streptococcus pyogenes has been shown to be a glyceraldehyde-3-phosphate dehydrogenase (GAPDH) (Pancholi & Fischetti, 1992). Furthermore, these cyto- solic enzymes are found on the cell surface. GAPDH is a major protective antigen found on the surface of Schistosoma mansoni (Goudot-Crozel et al., 1989). GAPDH also functions as a surface lectin responsible for flocculation of the yeast Kluyveromyces marxianus (Fernandes et al., 1992). Pertinent to C. albicans it has been shown that the immunodominant C. albicans glycolytic enzyme enolase is found in culture super- natants and on the surface of the fungus (Sundstrom & Aliaga, 1994), and the cytosolic enzyme 3-phos- phoglycerate kinase is located on the surface of the fungus (Alloush et al., 1997). GAPDH of C. albicans is cell-wall-associated and even binds fibronectin and laminin (Gozalbo et al., 1998). These examples of cytosolic enzymes performing in different capacities are perhaps examples of gene sharing, a concept best demonstrated in the vertebrate cornea and lens where the enzyme ADH serves as a major structural protein with little or no enzymic activity (Cooper et al., 1993).
