---
title: analyze TANGO result for XP_028889033 homologs
author: Bin He, basad on scripts written by Rachel Smoak
date: 2020-07-26
output:
    html_notebook:
        toc: yes
        toc_float: yes
---

```{r setup}
suppressPackageStartupMessages(library(Biostrings))  # for pairwise alignment and calculating distance
suppressPackageStartupMessages(library(tidyverse))   # for data wrangling and visualization
#suppressPackageStartupMessages(library(Hmisc))       # for summary functions
suppressPackageStartupMessages(library(cowplot))     # supplement ggplot
```

The libraries below are used for deriving and plotting weblogos
```{r install_library}
if(!requireNamespace("msa", quietly = TRUE))
  BiocManager::install("msa")
suppressPackageStartupMessages(library("msa"))

if(!requireNamespace("RWebLogo", quietly = TRUE))
  install.packages("RWebLogo")
suppressPackageStartupMessages(library("RWebLogo"))
```

## Goal

- Analyze TANGO predictions to extract $/beta$-aggregation prone sequences in XP_028889033 homologs.
- Attempt to cluster those sequences by amino acid sequence similarity, and explore the question of the evolution of the C-terminal portion (non Hyg_reg_CWP portion) of the protein homologs, from the perspective of the $\beta$-aggregation prone motifs. 

    In other words, we are curious about whether homologs that are more closely related based on the N-terminal domain also share more similarity in their C-terminus sequences. The expectation is yes, which would support the evolutionary model where the ancestral gene was repeatedly duplicated independently in various lineages, and the duplicates subsequently went through fast evolution of their C-terminus, potentially through de novo mutations and repeat-mediated tandem duplications and losses, which led to different lengths and composition of the C-terminus. This, in turn, made the homologs show different total length and domain architecture, which either reflected species-specific requirement for the same function or evolution of distinct functions. For the former, one could imagine that for a homolog in one species to function as an adhesin, the NTD must protrude outside the cell wall matrix, and since different species have different cell wall matrix thickness and composition, the stalk of the protein must adjust itself so the NTD is raised enough to protrude outside.
    
## Analysis
### Custom function to extract information from TANGO output

This function extracts the information of interest from the TANGO output file. It is designed to be used on a single file. Looping through all files in a folder can be easily achieved with `sapply` over a list of files using `list.files`.

```{r extract_tango_info}
extract_tango <- function(tango_output, agg_threshold = 5, required_in_serial = 5) {
    require(tidyverse)
    tmp <- read_tsv(file = tango_output, col_types = "icddddd") %>% 
        # a boolean vector for residues above threshold
        mutate(pass = Aggregation > agg_threshold)
    pass.rle <- rle(tmp$pass) # this creates a run length encoding that will be useful for identifying the sub-sequences in a run longer than certain length
    # --- Explanation ---
    # this rle object is at the core of this function
    # an example of the rle looks like
    #   lengths: int[1:10] 5 19 20 8 1 5 19 6 181 18
    #   values: logi[1:10] F T  F  T F T F  T F   T
    #   note that by definition the values will always be T/F interdigited
    # our goal is to identify the sub-sequences that is defined as a stretch of 
    # n consecutive positions with a score greater than the cutoff and record the
    # sub-sequence, its length, start and end position, 90% quantile of the score
    # --- End of explanation ---
    # 1. assigns a unique id for each run of events
    tmp$group <- rep(1:length(pass.rle$lengths), times = pass.rle$lengths)
    # 2. extract the subsequences
    agg.seq <- tmp %>% 
        filter(pass) %>% # drop residues not predicted to have aggregation potential
        group_by(group) %>% # cluster by the runs
        summarize(seq = paste0(aa, collapse = ""),
                  start = min(res), end = max(res), length = n(),
                  median = median(Aggregation),
                  q90 = quantile(Aggregation, probs = 0.9),
                  ivt = sum(aa %in% c("I","V","T")) / length(aa),
                  .groups = "drop") %>% 
        mutate(interval = start - lag(end) - 1) %>% 
        filter(length >= required_in_serial) %>% 
        select(-group)
    return(agg.seq)
}
```

### Apply to my XP_028889033 homologs
```{r apply}
getwd()
tango.output.files <- list.files(path = "tango-output", pattern = ".txt|.txt.gz", full.names = T)
# the read_csv() function used in the custom function can automatically decompress gzipped files
tango.res <- lapply(tango.output.files, extract_tango)
names(tango.res) <- gsub(".txt|.txt.gz", "", basename(tango.output.files))
# to add species information
seqInfo <- read_tsv("XP_028889033_homologs.tsv", comment = "#", col_types = c("ccci"))
tango.res.df <- bind_rows(tango.res, .id = "id") %>% 
  left_join(select(seqInfo, -length), by = c("id" = "id"))
# save the tango output
write_tsv(tango.res.df, "tango_summary_table.tsv.gz")
# mutate(species = str_split(id, "_(?!.*_)", simplify = TRUE)[,2]) 
# extract the species names
# credit: https://stackoverflow.com/questions/20454768/how-to-split-a-string-from-right-to-left-like-pythons-rsplit
# the split pattern is equivalent to the rsplit() function in python
```

### Exploratory data analysis of TANGO motifs
We will group the TANGO sequences into two groups based on whether they are in within the N-terminal Hyp_reg_CWP domain or not.
```{r}
# include the pfam domain information to classify the tango sequences
pfam <- read_tsv("raw-output/HMMER-HMMScan-Pfam-hits.tsv", col_types = "ciiiicciiidddiic")
tango.res.df1 <- tango.res.df %>% 
  left_join(pfam %>% 
              filter(hmm_name == "Hyphal_reg_CWP") %>% 
              select(name = seq_id, domain.start = envelope_start, domain.end = envelope_end), 
            by = c("name" = "name")) %>% 
  mutate(in.domain = (start >= domain.start-10 & end <= domain.end)) %>% 
  select(-domain.start, -domain.end)
# find unique motifs and count the number of proteins and species represented
motif.summary <- tango.res.df1 %>% 
    group_by(in.domain, seq) %>% 
    summarize(n = n(), n.prot = n_distinct(id), n.sps = n_distinct(species),
              medScore = round(mean(median),1),
              IVT = round(mean(ivt),2),
              avg.intv = round(mean(interval, na.rm = T),1), sd.intv = round(sd(interval, na.rm = T),1),
              species = paste0(unique(species), collapse = ","), .groups = "drop") %>% 
    arrange(desc(n))

# TANGO sequences inside the Hyphal_reg_CWP domain
print(filter(motif.summary, in.domain) %>% select(-in.domain))
# TANGO sequences outside the Hyphal_reg_CWP domain
print(filter(motif.summary, !in.domain) %>% select(-in.domain))
# what is the level of cross-species use of the same motif?
motif.summary %>% count(n.sps)
```
_**Discussion**_

1. TANGO sequences within the Hyphal_reg_CWP domain are distinct in their sequences from the ones outside and tend to be low in "IVT" content and low in beta-aggregations score.
1. Most of the motifs are species specific.

```{r eda, include=FALSE, fig.width=10, fig.height=4}
# Is there any relationship between the length of a motif and its aggregation value?
p0 <- motif.summary %>% 
    ggplot(aes(x = nchar(seq), y = avg.q90)) + geom_jitter(width = 0.2, alpha = 0.6) + 
    stat_summary(fun.data = mean_cl_boot, geom = "crossbar", col = "red") +
    labs(title = "Average of 90th percentile Beta aggregation prob. for TANGO motifs of different lengths")

p1 <- motif.summary %>% 
    ggplot(aes(x = nchar(seq), y = avg.median)) + geom_jitter(width = 0.2, alpha = 0.6) + 
    stat_summary(fun.data = mean_cl_boot, geom = "crossbar", col = "red") +
    labs(title = "Average of median of Beta aggregation prob. for TANGO motifs of different lengths")
p0
save_plot("img/20200727-motif-length-by-value-q90.png", p0, base_asp = 2)
p1
save_plot("img/20200727-motif-length-by-value-median.png", p1, base_asp = 2)

# Where are the TANGO identified motifs located within a sequence?
# color by 90% quantile of agg. probability
p2 <- tango.res.df %>% 
    mutate(score = cut(q90, breaks = c(0,25,50,75,100))) %>% 
    ggplot(aes(start, fill = score)) + geom_histogram(binwidth = 50) +
    scale_fill_brewer(type = "seq", palette = "YlGnBu")

p3 <- tango.res.df %>% 
    filter(start <= 1000) %>% 
    mutate(score = cut(q90, breaks = c(0,25,50,75,100))) %>% 
    ggplot(aes(start, fill = score)) + geom_histogram(binwidth = 25) +
    scale_fill_brewer(type = "seq", palette = "YlGnBu")

plot_grid(p2, p3)
ggsave("img/20200727-motif-location-group-by-agg-score.png")

# color by sequence motif length
p4 <- tango.res.df %>% 
    mutate(Length = cut(length, breaks = c(5,7,9,11,25), include.lowest = TRUE)) %>% 
    ggplot(aes(start, fill = Length)) + geom_histogram(binwidth = 50) +
    scale_fill_brewer(type = "seq", palette = "YlGnBu")

p5 <- tango.res.df %>% 
    filter(start <= 1000) %>% 
    mutate(Length = cut(length, breaks = c(5,7,9,11,25), include.lowest = TRUE)) %>% 
    ggplot(aes(start, fill = Length)) + geom_histogram(binwidth = 25) +
    scale_fill_brewer(type = "seq", palette = "YlGnBu")

plot_grid(p4, p5)
ggsave("img/20200727-motif-location-group-by-length.png")
```

### Cluster motifs by similarity
Many motifs are similar to each other but are listed differently, such as "GVVIVTT" and "GVIVVTT" etc. Therefore, we wonder if we can group them based on the similarity. Almost all clusterin algorithms are based on a pairwise distance matrix. To caluate the distance between any two sequences, we have the following consideration:
First, one can work with the Euclidian distance if the sequences are of the same length. Unfortunately, the TANGO predicted sequences are of different lengths, and thus we need a different measure. One such method is the edit distance, also called the [Levenshtein edit distance]()

### Levenshtein edit distance approach

Below are my first attempt, using the `stringdist` package.
```{r gather_motif}
library(stringdist)
# get the unique motif sets
seqs <- motif.summary %>% filter(!in.domain) %>% pull(seq)
# compute distance matrix between strings using the full Damerau-Levenshtein method
# reference: https://medium.com/my-datascience-journal/using-string-distance-to-group-text-variables-in-r-aa6d0f9226a6
motif.dist <- stringdistmatrix(seqs, method = "dl", weight = c(d = 0.2, i = 0.2, s = 0.1, t = 1), useNames = TRUE)
# by assigning different weights to deletion, insertion, substitution and transposition, I'm actively discouraging transposition, which cannot be achieved by any known mutation mechanism in one step. I'm also disfavoring insertion and deletion, although i would have loved to be able to distinguish between indels at the end (end gaps) from internal indels. The former is totally fine (I would like to cluster a substring of a longer motif with that longer one).
```

```{r test_clustering_method}
# test the performance of different hierarchical clustering method
# reference: https://uc-r.github.io/hc_clustering
library(cluster)
# methods to assess
m <- c( "average", "single", "complete", "weighted", "ward")
names(m) <- c( "average", "single", "complete", "weighted", "ward")

# function to compute coefficient
ac <- function(x) {
  agnes(motif.dist, method = x)$ac
}

map_dbl(m, ac)
```
- The ward method produces the strongest cluster.

```{r cluster, fig.width=10, fig.height=5}
motif.clust <- agnes(motif.dist, method = "ward")
pltree(motif.clust, cex = 0.07, hang = -1)
rect.hclust(as.hclust(motif.clust), h = 4, border = 1:10)
abline(h = 4, lty = 2)
dev.print(pdf, "img/20200727-motif-cluster-dendrogram.pdf")
```
_**Discussion**_

- Visually I can roughly detect 6 or 7 clusters. Next let's try to cut the tree into those clusters and examine the output

```{r explore_diff_cluster}
h <- seq(from = 3, to = 5, by = 0.5)
seq_grp <- lapply(h, function(x) cutree(as.hclust(motif.clust), h = x))
names(seq_grp) <- paste0("h", h)
cluster_bars <- do.call(cbind, seq_grp); 
pltree(motif.clust, cex = 0.05, hang = -1, xlab = "Tango sequences")
dendextend::colored_bars(cluster_bars, as.dendrogram(motif.clust), add = TRUE, y_shift = -3, cex.rowLabels = 0.8)
#abline(h = h, lty = 3)
dev.print(pdf, "img/20200825-tango-word-hierarchical-clustering.pdf", width = 12, height = 5)
```

_Discussion_

1. The first three clusters are very clear.
1. Most variability comes from the right half of the tree, where the tango sequences are longer.
1. Let's choose the lowest cutting height (3) and see what the clusters look like.

```{r}
motif.summary$seq_grp <- seq_grp$h3
# number of sequences clustered in the first 6 clusters
table(motif.summary$seq_grp)
# sequences belonging to cluster 1
toplist <- c("group 1" = 1, "group 2" = 2, "group 3" = 3)
lapply(toplist, function(x) paste(head(pull(filter(motif.summary, seq_grp == x), seq), n = 50), collapse = "  "))
```

_Discussion_

From the clustering result, I can identify the main issue with the currently approach -- because I set up the scoring matrix such that insertion and deletions are twice as costly as the substitutions, and since there is no way to distinguish the end gaps from the internal gaps, I'm stuck at either allowing internal gaps, which is not very desirable, or penalizing end gaps, which would result in sequences such as "GVVIV" and "VIVTT" to be grouped separately from the dominant sequence, "GVVIVTT". One can also see the effect of this by noticing how the tree is nicely ordered by the sequence motif length (left-to-right: short-to-long).

In an attempt to relieve this problem, I will decrease the gap penalty and see if it makes any difference.

### Levenshtein edit distance with different costs
```{r}
# reduce the insertion and deletion costs
motif.dist <- stringdistmatrix(seqs, method = "dl", weight = c(d = 0.12, i = 0.12, s = 0.1, t = 1), useNames = TRUE)
motif.clust <- agnes(motif.dist, method = "ward")
```

```{r}
h <- seq(from = 2.5, to = 5, by = 0.5)
seq_grp <- lapply(h, function(x) cutree(as.hclust(motif.clust), h = x))
names(seq_grp) <- paste0("h", h)
cluster_bars <- do.call(cbind, seq_grp); 
pltree(motif.clust, cex = 0.05, hang = -1, xlab = "Tango sequences")
dendextend::colored_bars(cluster_bars, as.dendrogram(motif.clust), add = TRUE, y_shift = -1, cex.rowLabels = 0.8)
```

```{r}
motif.summary$seq_grp <- seq_grp$h3
# number of sequences clustered in the first 6 clusters
table(motif.summary$seq_grp)
# sequences belonging to cluster 1
toplist <- c("group 1" = 1, "group 2" = 2, "group 3" = 3, "group 4" = 4, "group 5" = 5)
lapply(toplist, function(x) paste(head(pull(filter(motif.summary, seq_grp == x), seq), n = 50), collapse = "  "))
```
This result is still not satisfactory. To me, some sequences in group 3, e.g. "IVITT", "VIITT", "VVIVTT", "VIVTT", should be clustered with group 1, but they are not.

### K-medoids clustering, aka partition around medoids (PAM)
Let's try a different clustering approach, K-medoids, or Partition Around Medoids algorithm
For methods, refer to [wikipedia](https://en.wikipedia.org/wiki/K-medoids) and [r doc for the `pam()` function](https://www.rdocumentation.org/packages/cluster/versions/2.1.0/topics/pam). According to the latter, PAM has several advantages over K-means, including clustering around one of the objects (by miniziming within cluster dissimilarity _around_ that medoid). This has a great advantage for this case as I can easily examine the "meoids" after the clustering and see if those representatives should have been grouped together.
```{r silhouette_optimal_number_of_clusters}
# construct dissimilarity matrix
# we will use the distance matrix constructed above, i.e. motif.dist
# the following code is adapted from 
# https://www.rdocumentation.org/packages/cluster/versions/2.1.0/topics/pam.object
## Use the silhouette widths for assessing the best number of clusters,
## following a one-dimensional example from Christian Hennig :
##
asw <- numeric(25)
## Note that "k=1" won't work!
for (k in 2:25)
  asw[k] <- pam(motif.dist, k, diss = TRUE) $ silinfo $ avg.width
k.best <- which.max(asw)
cat("silhouette-optimal number of clusters:", k.best, "\n")

plot(1:25, asw, type= "b", pch = 19, frame = FALSE, main = "pam() clustering assessment",
     xlab= "k  (# clusters)", ylab = "average silhouette width")
axis(1, k.best, paste("best",k.best,sep="\n"), col = "red", col.axis = "red")
```

This is a bit surprising as the silhouette method suggests an optimal number of cluster = 2
```{r}
motif.pam <- pam(motif.dist, k = 10, diss = TRUE)
```

### Excluding motifs in the first 300 amino acids

```{r cluster_motif_1}
# remake the motif.summary table with those motifs in the N300 a.a. removed
motif.summary1 <- tango.res.df %>% 
    filter(start > 300) %>% 
    group_by(seq) %>% 
    summarize(n = n(), n.protein = n_distinct(id), n.species = n_distinct(species), 
              avg.q90 = round(mean(q90),1), sd.q90 = round(sd(q90),2),
              avg.intv = round(mean(interval, na.rm = T),1), sd.intv = round(sd(interval, na.rm = T),2),
              pc.N300 = sum(start <= 300)/n()*100,
              species = paste0(unique(species), collapse = ","), .groups = "drop") %>% 
    arrange(desc(n))
# get the unique motif sets
seqs1 <- motif.summary1$seq
# compute distance matrix between strings using the full Damerau-Levenshtein method
# reference: https://medium.com/my-datascience-journal/using-string-distance-to-group-text-variables-in-r-aa6d0f9226a6
motif.dist1 <- stringdistmatrix(seqs1, method = "dl", useNames = TRUE,
                                weight = c(d = 0.12, i = 0.12, s = 0.1, t = 1))
motif.clust1 <- agnes(motif.dist1, method = "ward")
pltree(motif.clust1, cex = 0.05, hang = -1)
dev.print(pdf, "img/20200727-motif-cluster-dendrogram-mN300.pdf")
```

```{r explore_diff_cluster_1}
motif.summary1$seq_grp <- cutree(as.hclust(motif.clust1), h = 2)
# number of sequences clustered in the first 6 clusters
head(table(motif.summary1$seq_grp))

# sequences belonging to cluster 1
toplist <- c("group 1" = 1, "group 2" = 2, "group 3" = 3)
lapply(toplist, function(x) paste(pull(filter(motif.summary1, seq_grp == x), seq), collapse = "  "))

# motif usage in each species
tango.res.df %>% 
    filter(start > 300) %>% 
    left_join(select(motif.summary, seq, seq_grp), by = c("seq" = "seq")) %>% 
    group_by(species) %>% 
    summarize(motif.use = paste(sort(unique(seq_grp)), collapse = ","))
```

### Using proper amino acid substitution matrix to calculate distance
Here we use the `Biostrings::stringDist()` function to calculate the pairwise distances between the motifs. What's different about this method compared to the generic string distance function used above is that it is based on pairwise alignment. Specifically, the function uses the following procedure to calculate the distance between any pair of sequences:

1. perform pairwise alignment (local/global) using a given substitution matrix (e.g. BLOSUM62)
1. use the negative of the score as the distance measure


    note that this is different from the edit distance (Levenshtein), which is what is used in the method above

Because the distance calculation is based on pairwise alignment, we need to decide on the "type" of alignment and the scoring matrix to use. According to the [manual](https://bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/PairwiseAlignments.pdf) of `Biostrings`, there are five types, which differ in which part of the sequence the substitution scoring and gap panelties are applied to:

1. Global

    end-to-end: $[a_1, b_1] = [1,n_1]$ and $[a_2,b_2] = [1,n_2]$, where $n_1$ and $n_2$ denote the lengths of the query and subject sequences, and $a_i$ and $b_i$ denote the starting and end position in the query and subject that the alignment scoring is applied to.

1. Local

    align string fragments: $[a_1, b_1]$ and $[a_2, b_2]$; based on [wikipedia](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm), the local alignment differs from the global alignment in that it sets the negative scoring matrix cells to zero. This means in this mode mismatches are _not penalized_.
    
1. Overlap

    align whole strings without **end gap** penalties: ${[a_1, b_1] = [a_1,n_1], [a_2, b_2] = [1, b_2]}$ or ${[a_1, b_1] = [1,b_1], [a_2, b_2] = [a_2, n_2]}$
    
1. Subject overlap (global-local)

    align whole strings in `pattern` with consecutive subsequences of `subject`: $[a_1,b_1]=[1,n_1]$ and $[a_2,b_2]$

1. Query overlap (local-global)

    align consecutive subsequences in `pattern` with whole strings of subject: $[a_1,b_1]$ and $[a_2,b_2]=[1,n_2]$
    
As for the scoring matrix, we will use a fixed one, BLOSUM62. The `pairwiseAlignment()` also allows for a quality-based scoring matrix. But since our sequences have no associated "quality", or "weights", we will not use that feature.

Lastly, the default for gap opening and extension penalty are 10 and 4, which are quite high for the length of the motif sequences. Therefore we should see no gapped alignment, which I think is the right choice.
```{r test_pairwise_alignmen_params}
# set.seed(seed = 123)
# sample.patterns <- sample(seqs, 10)
sample.patterns <- c("GVVIVTT", "FGVVTITT", "IVIITSGSGIVIVTT", "FTTFTTTFTT", "VLVTT", 
                     "GFFVTSL", "LIVFY", "LSLQALLSAVYCLSSVLALTIT", "FTSIWTTV", "FFSFWLLLNSVLALVIT")
mat <- "BLOSUM50"
pa1 <- pairwiseAlignment(sample.patterns, "GVVIVTT", type = "global", substitutionMatrix = mat)
pa2 <- pairwiseAlignment(sample.patterns, "GVVIVTT", type = "local", substitutionMatrix = mat)
pa3 <- pairwiseAlignment(sample.patterns, "GVVIVTT", type = "overlap", substitutionMatrix = mat)
data.frame(pattern = sample.patterns, glb.aln = aligned(pa1), glb.d = nedit(pa1), glb.S = score(pa1),
           lc.aln = aligned(pa2), lc.d = nedit(pa2), lc.S = score(pa2),
           ov.aln = aligned(pa3), ov.d = nedit(pa3), ov.S = score(pa3))
```

Based on the result above, I decide to try "overlap" (which requires at least one end be anchored) with "BLOSUM" substitution matrix to calculate the distance
```{r}
motif.dist2 <- stringDist(seqs, method = "substitutionMatrix", type = "overlap", substitutionMatrix = "BLOSUM50", gapOpening = 10, gapExtension = 4)
```
```{r determine_optimal_num_clusters}
asw <- numeric(20)
## Note that "k=1" won't work!
for (k in 2:20)
  asw[k] <- pam(motif.dist2, k, diss = TRUE) $ silinfo $ avg.width
k.best <- which.max(asw)
cat("silhouette-optimal number of clusters:", k.best, "\n")

plot(1:20, asw, type= "b", pch = 19, frame = FALSE, main = "pam() clustering assessment",
     xlab= "k  (# clusters)", ylab = "average silhouette width")
axis(1, k.best, paste("best",k.best,sep="\n"), col = "red", col.axis = "red")
```

```{r}
motif.clust2 <- agnes(-motif.dist2, method = "ward")
pltree(motif.clust2, cex = 0.05, hang = -1)
```

