# Paper Outline
## 16 November 2020
Suggested paper outline:
### Introduction
	New species, multi-drug resistant, very sticky
		Adhesins implicated in these things
	Adhesins unknown
		No good homologs to well-characterized species
		Talk about how we disagree with the annotations out there
	Adhesin predictive programs
		Nothing trained on any similar species, so even though they can be used can they really identify novel adhesins?
		Even known adhesin structures have large differences
	Our approach: look for traits associated with adhesins and find a putative adhesin protein family
### Methods/Results/Discussion
#### Case study
		How do we pick Lindsay's protein as the starting point?
			Look at Bin's case study presentation
		Sequence features:
			Domain architecture, serine/threonine richness, aggretation sequences, GPI anchor, hydrophobic regions?,
      length, tandem repeats?, glycosylation sites?, chromosomal location?
		Structural features:
			ITASSER, CATH?
#### Homolog tree
		Homologs in MDR clade look similar to each other but not really like anyone else - still have features of adhesins
		ITASSER results for selected homologs - other adhesive structures, even though they don't map to the same template
			Still a bit odd to me because they reflect the same domain architecture?
		Section on GVVIVTT repeat and its potential function?
    
#### Expansion to other proteins that "look like" this group in C. auris
		Rank and suggest for bench-scale testing
		Compare to FRV and FAA results - would it find these putative adhesins? 
    Do we need a different type of tool/different training sets?
*Is there a place here to call for sequencing - the new/old glabrata information? Maybe
