---
title: "PhyloGLM analysis"
author: "Bin He"
date: "2022-08-18 (updated: `r Sys.Date()`)"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    toc_depth: 5
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages(require(tidyverse))
suppressPackageStartupMessages(require(cowplot))
suppressPackageStartupMessages(require(phytools))
suppressPackageStartupMessages(require(phylolm))
```

## Introduction
_Goal_

- Test the hypothesis that pathogenic Candida lineages have a larger Hil family, while controlling for phylogenetic relatedness

_Input_

- Species tree
- Gene family size vector

_Approach_

- t-test under the incorrect assumption of independence among the Hil family size values across species (they are related by the tree and thus are not i.i.d.)
- PhyloGLM using the `phylolm` package in R, which tests for association between a quantitative independent variable (Hil family size) and a binary dependent variable (pathogen or not)

## Analysis
### Load data
Species tree
```{r}
spsInfo <- read_tsv("../input/20220518-expanded-blast-species-info.tsv", col_types = cols())
tree <- read.tree("../input/20220521-generax-species-tree.nwk")
```

Encode the life history trait
```{r}
trait <- spsInfo %>% 
  mutate(species = treeName, 
         pathogen = fct_recode(pathogen, Yes = "human", Yes = "crustacean", No = "human (rare)", No = "no report"),
         .keep = "none")
```

Gene family size
```{r}
famSize <- read_tsv("../input/20220818-Hil-family-size-adhesin-status-summary.tsv", col_types = cols()) %>% 
  mutate(species = factor(gsub(" ", "_", species), levels = tree$tip.label)) %>% 
  group_by(species) %>% 
  summarize(total = n(), FRV = sum(frv.pred), SP = sum(sp.pred), GPI = sum(gpi.pred), 
            final = sum(frv.pred & sp.pred & gpi.pred)) %>% 
  select(species, total, final) %>% 
  # complete missing values for S. cerevisiae
  complete(species, fill = list(total = 0, final = 0))
```

### Visually compare the pathogen vs non-pathogens
```{r}
dat <- left_join(trait, famSize, by = "species")
```

We remove _M. bicuspidata_ here as it doesn't fit either the human pathogen or the low pathogenic potential group.
```{r}
dat1 <- dat %>% 
  filter(species != "Metschnikowia_bicuspidata")
```

Plot as boxplots
```{r}
p <- dat1 %>% 
  pivot_longer(c(total, final), names_to = "var", values_to = "n") %>% 
  mutate(var = fct_recode(var, Total = "total", `Putative adhesin` = "final") %>% 
           fct_relevel("Total")) %>% 
  ggplot(aes(x = pathogen, y = n)) + 
  geom_boxplot(width = 0.5) + 
  geom_jitter(width = 0.2, alpha = 0.7, size = 2, shape = 16) +
  ylab("# of Hil family genes") 
p + facet_wrap(~var) + theme_cowplot(font_size = 18)
ggsave(file = "../output/img/20220819-Hil-family-size-between-PA-nonPA.png", 
       plot = p + facet_wrap(~var, nrow = 2) + theme_bw(base_size = 12) + background_grid(major = "y"),
       width = 3, height = 4)
```

### Star phylogeny test
Here we make the incorrect assumption that all species are equally related to each other, such that both their traits and their family sizes evolved independently from the common ancestor. We are looking to test whether there is an association between the life history trait and the Hil family size

1. t-test for total number of Hil family homologs
```{r}
t.test(total ~ pathogen, data = dat1, alternative = "greater")
```

1. t-test for putative adhesins within the Hil family
```{r}
t.test(final ~ pathogen, data = dat1, alternative = "greater")
```

1. non-parametric rank sum test for total number of Hil family homologs
```{r}
wilcox.test(total ~ pathogen, data = dat1, paired = FALSE, alternative = "greater")
```
1. non-parametric rank sum test for putative adhesins within the Hil family
```{r}
wilcox.test(final ~ pathogen, data = dat1, paired = FALSE, alternative = "greater")
```

### Phylogenetic regression using PhyloGLM
Prepare data
```{r}
df <- dat %>% column_to_rownames(var = "species") %>% 
  mutate(pathogen = 2 - as.numeric(pathogen))
df1 <- dat1 %>% column_to_rownames(var = "species") %>% 
  mutate(pathogen = 2 - as.numeric(pathogen))
```

1. Fit the model with _M. bicuspidata_ included, using the total number of Hil family homologs as the independent variable
```{r}
fit = list()
fit$'all.total' <- phyloglm(formula = pathogen ~ total, data = df, phy = tree, method = "logistic_IG10", btol = 50, boot = 100)
summary(fit$"all.total")
```

1. Fit the model with _M. bicuspidata_ included, using the number of putative adhesins in the Hil family as the independent variable
```{r}
fit$'all.final' <- phyloglm(formula = pathogen ~ final, data = df, phy = tree, method = "logistic_IG10", btol = 50, boot = 100)
summary(fit$'all.final')
```

1. Fit the model without _M. bicuspidata_, using the total number of Hil family homologs as the independent variable
```{r}
tree1 <- drop.tip(phy = tree, tip = "Metschnikowia_bicuspidata")
fit$'noM.total' <- phyloglm(formula = pathogen ~ total, data = df1, phy = tree1, method = "logistic_IG10", btol = 50, boot = 100)
summary(fit$"noM.total")
```

1. Fit the model without _M. bicuspidata_, using the number of putative adhesins in the Hil family as the independent variable
```{r}
fit$'noM.final' <- phyloglm(formula = pathogen ~ final, data = df1, phy = tree1, method = "logistic_IG10", btol = 50, boot = 100)
summary(fit$'noM.final')
```