
1. Used fungiDB blast tool to identify homologs of my protein (filtered Saccharomycetes, default parameters except low complexity filter was on)
1. Took those 50 proteins and aligned them with clustalO 
1. Visualized alignment in Jalview and removed problematic proteins 
1. Realigned and cut the alignment at 430 amino acids to encompass the N terminal domain (where alignment started getting bad)
1. Built ML gene tree using MegaX (default parameters, bootstrap 50)
2. Uploaded gene and species tree to Notung
3. Reconciled and used rooting and rearranging tabs to create the final tree
